Add Tests to Presenter

TTitcombe opened the pull request:
### Description of work
This adds unit tests to the main window presenter, which was missing from the PR which introduced the gui

Currently, the mocked classes (mainwindow and IEX) do not use trompeloeil, as I couldn't get trompeloeil to work properly with pointers

This PR also turn MainWindow into a shared_ptr

### Issue

*No associated issue*

### Acceptance Criteria
1. Unit tests pass
2. Code review

### Unit Tests
`tests/UI/presenterTest.cpp`

---

## Code Review (To be filled in by the reviewer only)

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?

---

TTitcombe committed:
Presenter takes IEX class as a pointer to an IEXInterface

created IEXInterface. IEX derives from this. Allows us to mock out IEX

TTitcombe committed:
Create presenter test

TTitcombe committed:
Create presenter with mocked out view and iex

TTitcombe committed:
Successful unit tests for onStockButtonClicked method in presenter

View mocking does not use trompeloeil currently

TTitcombe committed:
Turn mainwindow into shared ptr

TTitcombe committed:
Add QtGui to CMakelists

catch reference to spdlogger exceptions

change app to qapp

matthew-d-jones reviewed (changes requested):
I haven't look through this in detail yet, but there are a couple of changes you need to make for gcc to be happy:
1. In `presenterTest` it doesn't like you using a string literal as a `char *` argument, this is because a string literal is `const`. Instead you could define a variable for the program name like this:
```cpp
char programName[] = "program name";
char *argv[] = {programName, nullptr};
```
Also note use of `nullptr` rather than `NULL`. `nullptr` should always be used instead of `NULL`. It is a pointer type so it is safer to use anywhere type deduction is involved, `NULL` can be treated as an integer leading to ambiguity in overload selection and that kind of thing.

2. The order of initialisation in the constructor should match the order of declaration in the class. gcc otherwise gives a warning and you have told it to treat warnings as errors. The constructor for `MainWindowMock` should be:
```cpp
explicit MainWindowMock(const std::string &mock_symbol, QWidget *parent = nullptr) :
    UI::MainWindow(parent), m_mock_symbol(mock_symbol) {}
```

TTitcombe committed:
Act on reviewer comments

Use string instead of string literal for main argc

change order of variable initialisation
